feat(boards): add nrf52 nosd snippets (#2545)
zmkfirmware/zmk: docs/docs/config/system.md

| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_STACK_SIZE`            | int  | Stack size of the BLE split peripheral notify thread                       | 650                                        |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_PRIORITY`              | int  | Priority of the BLE split peripheral notify thread                         | 5                                          |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_POSITION_QUEUE_SIZE`   | int  | Max number of key state events to queue to send to the central             | 10                                         |

## Snippets

Snippets are a way to save common configuration separately when it applies to multiple different applications.

> More documentation is avaialable from [Zephyr](https://docs.zephyrproject.org/latest/build/snippets/index.html).

Reviewer: ```suggestion
More documentation is avaialable from [Zephyr](https://docs.zephyrproject.org/3.5.0/build/snippets/index.html).
```
We pin these and update on Zephyr upgrades so we don't have dead links in the future.

Author: Done.

> More documentation is avaialable from [Zephyr](https://docs.zephyrproject.org/latest/build/snippets/index.html).

Reviewer 2: ```suggestion
More documentation is available from [Zephyr](https://docs.zephyrproject.org/3.5.0/build/snippets/index.html).
```

Author: Done.

> More documentation is avaialable from [Zephyr](https://docs.zephyrproject.org/latest/build/snippets/index.html).

Reviewer 2: Honestly, this text or something similar seems like it's more suited for somewhere in the configuration index, with this part being just a short link to somewhere. I'm fine with leaving it as-is for the time being, though.

> ZMK implements the following snippets:

Reviewer 2: ```suggestion
```
I wouldn't consider this to be a general "Snippets" section, rather just "These are the snippets that fall under system configuration".

Author: The issue here is there is no general "Snippets" section, this should be moved entirely in the future.

Reviewer 2: I'd say "Snippets" is fine, but I'd also be happy with "Memory Configuration".

Author: I updated the line to `ZMK implements the following system configuration snippets:` instead, this should hopefully remove the implication this is the only snippet ZMK adds.

### nrf52833-nosd

> 

Reviewer 2: I'd add a link to the snippet definition, just like how we do for devicetree stuff.

Author: Done.

On memory-constrained nRF52833 boards this snippet will extend the code partition to overwrite the Nordic SoftDevice.
This gives 428KB for the code partition as opposed to 280KB with the Nordic SoftDevice.

> The added memory allows the nRF52833 to still fit displays and other memory-intensive features.

Reviewer 2: ```suggestion
The added memory allows the nRF52833 to fit displays and other memory-intensive features.
```
Sounds nicer, imo.

> Enable the snippet with`-S nrf52833-nosd` for local builds, for example:
> 
> ```
> west build -b nrfmicro_13_52833 -S nrf52833-nosd -- -DSHIELD=corne_left
> ```
> 
> Or by adding `snippet: nrf52833-nosd` to your `build.yaml` for the appropriate board:
> 
> ```
>   - board: nrfmicro_13_52833
>     snippet: nrf52833-nosd
>     shield: corne_left
>     cmake-args: -DCONFIG_ZMK_DISPLAY=y

Reviewer: ````suggestion
Enable the snippet with `-S nrf52833-nosd` for local builds, for example:

```sh
west build -b nrfmicro_13_52833 -S nrf52833-nosd -- -DSHIELD=corne_left
```

Or by adding `snippet: nrf52833-nosd` to your `build.yaml` for the appropriate board:

```yaml
  - board: nrfmicro_13_52833
    shield: corne_left
    snippet: nrf52833-nosd
````
I'd move these up out of the subsection so that it is generic instructions for snippets, rather than for this one specifically.

(I am guessing you added the display arg related to the explanation above but we'd rather omit the potentially confusing extra thing here.)

Author: Moved up, added language tags.

>     cmake-args: -DCONFIG_ZMK_DISPLAY=y

Reviewer 2: Also information which should go under configuration (and building and flashing locally, for that part), but again I'm fine with it where it is for now to avoid overcomplicating this PR.

```

> :::warning
> Erasing the SoftDevice will not allow the board to use firmware built without the snippet.
> 
> Flashing such firmware WILL brick the board.
> The only way to restore functionality after that is to re-flash the bootloader.
> :::

Reviewer: ```suggestion
:::caution
Erasing the SoftDevice will prevent the board from using firmware built without the snippet.

Flashing such firmware **will** brick the board.
The only way to restore functionality after that is to re-flash the bootloader with one that contains the appropriate SoftDevice partition.
:::
```
I think this deserves the big red box 😄

Author: Added a big red box and additional clarifications.

```

For local builds, add `-S <snippet>` to your build command. For example:

```sh
west build -b nrfmicro_13_52833 -S nrf52833-nosd -- -DSHIELD=corne_left
```

ZMK implements the following system configuration snippets:

### nrf52833-nosd

Definition: [zmk/app/snippets/nrf52833-nosd](https://github.com/zmkfirmware/zmk/blob/main/app/snippets/nrf52833-nosd)

On memory-constrained nRF52833 boards this snippet will extend the code partition to overwrite the Nordic SoftDevice.
This gives 428KB for the code partition as opposed to 280KB with the Nordic SoftDevice.

The added memory allows the nRF52833 to fit displays and other memory-intensive features.

### nrf52840-nosd

Definition: [zmk/app/snippets/nrf52840-nosd](https://github.com/zmkfirmware/zmk/blob/main/app/snippets/nrf52840-nosd)

On nRF52840 boards this snippet will overwrite the Nordic SoftDevice, extending both the code and storage partitions.
This gives 844KB/128KB for the code/storage partitions as opposed to 792KB/32KB with the Nordic SoftDevice.

Firmware built with this snippet can work on boards after accidentally erasing the SoftDevice.
It can also be useful for especially memory-intensive applications.

> :::danger

Reviewer 3: IMHO, this warning should come first/early so as not to be skipped/missed.

| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_STACK_SIZE`            | int  | Stack size of the BLE split peripheral notify thread                       | 650                                        |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_PRIORITY`              | int  | Priority of the BLE split peripheral notify thread                         | 5                                          |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_POSITION_QUEUE_SIZE`   | int  | Max number of key state events to queue to send to the central             | 10                                         |

## Snippets

> Snippets are a way to save common configuration separately when it applies to multiple different applications.
> 
> More documentation is available from [Zephyr](https://docs.zephyrproject.org/3.5.0/build/snippets/index.html).

Reviewer 2: ```suggestion
[Snippets](https://docs.zephyrproject.org/3.5.0/build/snippets/index.html) are a way to save common configuration separately when it applies to multiple different applications.
```

Author: Done.

> Enable snippets with `-S <snippet>` for local builds, for example:
> 
> ```sh
> west build -b nrfmicro_13_52833 -S nrf52833-nosd -- -DSHIELD=corne_left
> ```
> 
> Or by adding `snippet: <snippet>` to your `build.yaml` for the appropriate board:
> 
> ```yaml
> - board: nrfmicro_13_52833
>   snippet: nrf52833-nosd
>   shield: corne_left
> ```

Reviewer 2: ````suggestion
Enable snippets by adding `snippet: <snippet>` to your `build.yaml` for the appropriate board:

```yaml
- board: nrfmicro_13_52833
  snippet: nrf52833-nosd
  shield: corne_left
```

For local builds, add `-S <snippet>` to your build command. For example:

```sh
west build -b nrfmicro_13_52833 -S nrf52833-nosd -- -DSHIELD=corne_left
```
````
Swapped order because we tend to recommend GHA first.

Author: Swapped, thank you. I tend to focus on local builds for development.

ZMK implements the following system configuration snippets:

### nrf52833-nosd

Definition: [zmk/app/snippets/nrf52833-nosd](https://github.com/zmkfirmware/zmk/blob/main/app/snippets/nrf52833-nosd)

> On memory-constrained nRF52833 boards this snippet will extend the code partition to overwrite the Nordic SoftDevice.
> This gives 428KB for the code partition as opposed to 280KB with the Nordic SoftDevice.
> 
> The added memory allows the nRF52833 to fit displays and other memory-intensive features.
> 
> ### nrf52840-nosd
> 
> Definition: [zmk/app/snippets/nrf52840-nosd](https://github.com/zmkfirmware/zmk/blob/main/app/snippets/nrf52840-nosd)
> 
> On nRF52840 boards this snippet will overwrite the Nordic SoftDevice, extending both the code and storage partitions.
> This gives 844KB/128KB for the code/storage partitions as opposed to 792KB/32KB with the Nordic SoftDevice.
> 
> Firmware built with this snippet can work on boards after accidentally erasing the SoftDevice.
> It can also be useful for especially memory-intensive applications.

Reviewer 2: ```suggestion
On nRF52833 boards this snippet will extend the code partition to overwrite the Nordic SoftDevice.
This gives 428KB for the code partition as opposed to 280KB with the Nordic SoftDevice.

The added memory allows the nRF52833 to fit displays and other memory-intensive features.

### nrf52840-nosd

Definition: [zmk/app/snippets/nrf52840-nosd](https://github.com/zmkfirmware/zmk/blob/main/app/snippets/nrf52840-nosd)

On nRF52840 boards this snippet will overwrite the Nordic SoftDevice, extending both the code and storage partitions.
This gives 844KB/128KB for the code/storage partitions as opposed to 792KB/32KB with the Nordic SoftDevice.

Firmware built with this snippet can work on boards after accidentally erasing the SoftDevice.
It can also be useful for especially memory-intensive applications.
```
Bit of improved parity.

Author: This bit is there to underline a lack of parity, though - an nRF52840 is generally fine, I haven't been able to run out of memory on any of mine yet, while a split central with an nRF52833 won't even build with displays without optimizations.

Reviewer 2: I feel that the last sentence in that section suffices on that, but I'm happy with keeping this bit in too.
